add support for multiple vault secrets engines

kevin-thompson opened the pull request:
To address the issue #1270 

Changes the property backend to backends and makes it a string array. Leaves "secret" as the default.
Updates the name for the property source to be "vault/{backend}:{key}" instead of just "vault:{key}"

pivotal-issuemaster commented:
@kevin-thompson Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-config&pullRequestId=1273)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-config&pullRequestId=1273) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

spencergibb commented on spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/VaultEnvironmentProperties.java:
```diff
@@ -102,12 +103,12 @@ public void setScheme(String scheme) {
 		this.scheme = scheme;
 	}
 
-	public String getBackend() {
```
There is a process to deprecate properties that spring boot has. We'll need to use it here. I'm on mobile, but if you haven't found it by Monday, ping me and I'll find it.

kevin-thompson replied:
I'll look into it, thanks.

pivotal-issuemaster commented:
@kevin-thompson Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-config&pullRequestId=1273)!

kevin-thompson commented:
> 
> 
> So you'll need to keep the old tests and add new as well to cover the deprecation

I didn't remove any test. I added a single test and cleaned up the mocking of the responses.

I did update the test that was using setBackend() to use setBackends(). Are you saying I should duplicate that with the one and continue to use the deprecated method in one place and use the new method in the other?

spencergibb commented:
Yes, are set back end if restored there needs to be tests that verify backwards compatibility

kamalakarp commented:
@kevin-thompson - the changes you did will only work for multiple KV(Key/Value) secret backends and it doesn't support database or other secret engines because the path is different.

KV secret engine: **vault read secret/configclient** which makes the URL : http://127.0.01:8200/v1/secret/foo

Database secret engine : **vault read database/creds/test-mongo-db-role** which makes the URL : http://127.0.01:8200/v1/database/creds/test-mongo-db-role

If your intention is to support the multiple secret backends like KV, database, and transit - then paths need to be adjusted properly and read the data from Vault

![image](https://user-images.githubusercontent.com/24970140/53677633-dd6a8b00-3c78-11e9-913c-844ad8b4cf43.png)

In the picture, config, cubbyhole, kv, and secret are variants of KV secret engines where as database and transit secret engines have different pathing.

kevin-thompson commented:
@kamalakarp 
Well, you can massage the correct path into place by setting a default key, but I agree that's not ideal.
The bigger problem with database backends is that they require a completely different response wrapper. Additionally, they necessitate (not always, but practically), the need to handle the lease information. That brings up a whole different issue because it's not the config server that's using the lease information, it's the consuming app. We could take care of the leasing on the config server side, but that would require a similar implementation of push notifications that the Git implementations use. I'm not prepared to tackle that here.

You bring up a great point about the kvVersion possibly applying differently, and I would not be opposed to reworking the code to turning backends into a list of objects rather than strings. This would better support the (hopefully) eventual support of the database backends (it would need additional properties such as the names of the properties into which the username/password should be dumped).

Also, you bring up the transit backend. I'm not sure that really belongs in the config server as it's just vault encrypting passed information, not setting up relatively static properties.

kevin-thompson commented:
@spencergibb I did not find any documentation describing how to deprecate properties, but I did find the annotation that I applied (@DeprecatedConfigurationProperty). That seems like what you were talking about.
@ryanjbaxter I updated the documentation. Let me know if I missed anything.

I'd like to hear back from @kamalakarp to get a response on the string list vs. object list thing.

kamalakarp commented:
@kevin-thompson - it would be better if we can make backend as a list of objects. 
With these changes, we will have to deprecate backend, defaultKey and kv.Version and make them as an object.

naxhh commented:
I've just started familiarizing myself with the project so dont take me so serious.

I agree with @kamalakarp in the fact that would be better to manage vault backends as a list of objects.
The current approach doesn't allow you to have KV backends with v1 and v2 at the same time.

Also the backends can be under any path, so that's also it will need to be taken care of.

Another thing I was thinking about is that some backends (like db, ones) have a ttl for the credentials, and you may need to renew those, I don't see how this is managed here, it's because the code will do polling of the resources? again not very familiar with the code yet.

The `.backend` `.kvVersion` and other configurations could be used as the first backend of the list if exist

spencergibb commented:
It feels like this is on the right path, but not supporting other backend types seems like missing something basic. I'm going to close this and we can revisit if there is any interest in picking this up.
